Splicing support

ksedgwic opened the pull request:
Upstream CLN support for VLS

ksedgwic committed:
Only run splicing tests when EXPERIMENTAL_SPLICING

Revert "splicing: Skip splice test until splicing is supported"

This reverts commit bd9494c65be09e6f8ecb3bc49b50a859467a65fe.

ksedgwic committed:
THROWAWAY: Add print debugging to test_splice

ksedgwic committed:
tests: Skip unsupported tests

ksedgwic committed:
THROWAWAY: avoided a unit test link issue

ksedgwic committed:
WIP: relax psbt_has_required_fields, psbt debugging

devrandom committed:
WORKAROUND: disable psbt_contribs_changed check because it deletes our signatures

ksedgwic committed:
tests: skip sign_option_will_fund_offer because (VLS:[#399])

devrandom commented on channeld/channeld.c:
```diff
@@ -3450,6 +3450,39 @@ static struct inflight *inflights_new(struct peer *peer)
+		local_upfront_shutdown_script,
+		local_upfront_shutdown_wallet_index,
+		&peer->channel->basepoints[REMOTE],
+		&peer->channel->funding_pubkey[REMOTE],
```
the funding pubkey is allowed to vary in the splice, and we need to be involved in that.  I'm wondering where that happens and how we will be informed of the remote one (and asked about the local one).

ksedgwic replied:
> we need to be involved in that

Are you saying that roundtrips/protocol is missing?

> how we will be informed of the remote one

I think we are being informed with this call.  Am I missing something?

ksedgwic replied:
> funding pubkey is allowed to vary in the splice

https://github.com/lightning-signer/c-lightning/blob/2023-08-explore-splicing/channeld/channeld.c#L3536-L3539

That seems to imply it is not changing?

ksedgwic replied:
This is where we are telling the other side our key:

https://github.com/lightning-signer/c-lightning/blob/2023-08-explore-splicing/channeld/channeld.c#L3548-L3552

ksedgwic replied:
each of the prior is one side, presume there are similar in the other "direction"

devrandom replied:
the spec allows for it to change (for the whole splice, not per RBF candidate).  we should ask the CLN team about this.

ksedgwic replied:
I just scoured the spec looking for where it is allowed to change ... can you point me to it?

devrandom replied:
https://github.com/lightning/bolts/pull/863/files#diff-ed04ca2c673fd6aabde69389511fa9ee60cb44d6b2ef6c88b549ffaa753d6afeR522

```
    type: 74 (splice)

    data:
        [chain_hash:chain_hash]
        [channel_id:channel_id]
        [u64:funding_satoshis]
        [u32:funding_feerate_perkw]
        [u32:locktime]
        [point:funding_pubkey]

    type: 76 (splice_ack)

    data:
        [chain_hash:chain_hash]
        [channel_id:channel_id]
        [u64:funding_satoshis]
        [point:funding_pubkey]
```

devrandom replied:
looks like CLN explicitly doesn't support changing them, but I assume that will be fixed before release.  need to ask them.

devrandom commented:
the location of the calls looks reasonable, but a couple of open questions above

ksedgwic committed:
hsmd: Rename hsmd_ready_channel to hsmd_setup_channel

When BOLT-2 renamed funding_locked to channel_ready the hsmd message
named ready_channel became very confusing.  Change it to setup_channel
instead to not confuse developers.not trigger the wrong association.

Because the hsmd wire messages are parsed by number this change is
forward and backwards compatible.

Changelog-Changed: hsmd: Renamed hsmd_ready_channel to hsmd_setup_channel because of the BOLT-2 rename of funding_locked to channel_ready

ksedgwic committed:
splice: Add hsmd_check_outpoint and hsmd_lock_outpoint ([#6722])

Changelog-Added: Added hsmd_check_outpoint and hsmd_lock_outpoint per ([#6722])

ksedgwic committed:
splice: Add hsmd_next_funding_pubkey

ksedgwic committed:
splice: Add call to hsmd_{check,lock}_outpoint on mutual splice_locked

ksedgwic committed:
splice: Add call to hsmd_{check,lock}_outpoint on mutual channel_ready

vincenzopalazzo commented on channeld/channeld.c:
```diff
@@ -3586,6 +3619,8 @@ static void splice_accepter(struct peer *peer, const u8 *inmsg)
 	new_inflight->last_tx = NULL;
 	new_inflight->i_am_initiator = false;
 
+	update_hsmd_with_splice(peer, new_inflight);
```
So in this case the amount that the hsmd needs to know is the total amount, right?

We can calculate by the `current funding tx amount + remote one`.

vincenzopalazzo replied:
That sounds good to me, just to rephrase in other words, the signer with the `setup_channel` needs to know the current total amount of the channel and then also 
a relative amount (`push_val`) that we are adding with the splice.

This looks good to me.

I also noted that I made a typo in my previous message, the formula that I was suggesting is `current funding tx amount + relative amount`

ksedgwic committed:
splice: signer must be informed of splice params

The signer needs to know when the splice operation starts and the
splice parameters for each splice transaction candidate.

The channel establishment v2 (dual funding) code path already
notifies the signer via the hsmd API hsmd_ready_channel calls
However, the splicing code path does not.

Link: https://github.com/ElementsProject/lightning/issues/6723
Suggested-by: @devrandom
Co-Developed-by: @devrandom
Co-Developed-by: Ken Sedgwick <ken@bonsai.com>
Signed-off-by: Vincenzo Palazzo <vincenzopalazzodev@gmail.com>

vincenzopalazzo commented on hsmd/hsmd_wire.csv:
```diff
@@ -89,6 +89,17 @@ msgdata,hsmd_setup_channel,channel_type,channel_type,
 # No value returned.
 msgtype,hsmd_setup_channel_reply,131
 
+# Sent to derive the next funding pubkey (for splicing)
+msgtype,hsmd_next_funding_pubkey,34
+msgdata,hsmd_next_funding_pubkey,peerid,node_id,
+msgdata,hsmd_next_funding_pubkey,dbid,u64,
+msgdata,hsmd_next_funding_pubkey,funding_txid,bitcoin_txid,
+msgdata,hsmd_next_funding_pubkey,funding_txout,u32,
```
I found this while cherripick the commits, we need it in cln? or it is just an experiment in vls?

devrandom replied:
you will need it once CLN wants to support funding pubkey rotation, but that would be a reasonably big change and is not needed right away (can keep the key unrotated with some loss of privacy).

vincenzopalazzo replied:
oh nice thanks! so will drop it for now
